listenerutil/parse.go

```diff
@@ -387,3 +413,122 @@ func ParseSingleIPTemplate(ipTmpl string) (string, error) {
 		return "", fmt.Errorf("multiple addresses found (%q), please configure one", out)
 	}
 }
+
+var validCustomStatusCodeCollection = []string{
+	"default",
+	"1xx",
+	"2xx",
+	"3xx",
+	"4xx",
+	"5xx",
+}
+
+const strictTransportSecurity = "max-age=31536000; includeSubDomains"
+
+// parseCustomResponseHeaders takes a raw config values for the
+// "custom_response_headers". It makes sure the config entry is passed in
+// as a map of status code to a map of header name and header values. It
+// verifies the validity of the status codes, and header values. It also
+// adds the default headers values.
+func parseCustomResponseHeaders(responseHeaders interface{}) (map[string]map[string]string, error) {
+	h := make(map[string]map[string]string)
+	// if r is nil, we still should set the default custom headers
+	if responseHeaders == nil {
+		h["default"] = map[string]string{"Strict-Transport-Security": strictTransportSecurity}
+		return h, nil
+	}
+
+	customResponseHeader, ok := responseHeaders.([]map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("response headers were not configured correctly. please make sure they're in a slice of maps")
+	}
+
+	for _, crh := range customResponseHeader {
+		for statusCode, responseHeader := range crh {
+			headerValList, ok := responseHeader.([]map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("response headers were not configured correctly. please make sure they're in a slice of maps")
+			}
+
+			if !isValidStatusCode(statusCode) {
+				return nil, fmt.Errorf("invalid status code found in the server configuration: %v", statusCode)
+			}
+
+			if len(headerValList) != 1 {
+				return nil, fmt.Errorf("invalid number of response headers exist")
+			}
+			headerValMap := headerValList[0]
+			headerVal, err := parseHeaders(headerValMap)
+			if err != nil {
+				return nil, err
+			}
+
+			h[statusCode] = headerVal
+		}
+	}
+
+	// setting Strict-Transport-Security as a default header
+	if h["default"] == nil {
+		h["default"] = make(map[string]string)
+	}
+	if _, ok := h["default"]["Strict-Transport-Security"]; !ok {
+		h["default"]["Strict-Transport-Security"] = strictTransportSecurity
+	}
+
+	return h, nil
+}
+
+// isValidStatusCode checking for status codes outside the boundary
```

Comment:
This could be a tiny bit confusing 😂.
```suggestion
// isValidStatusCode checks for status codes outside the allowed range
```